### ISSUE TITLE
Allow creating empty patterns from Lua

### ISSUE THREAD

**Issue by TheDJVG** [haproxy/haproxy#2202]
### Your Feature Request

Allow the Lua `core.register_init()` function to create instance of Map (something like: `Map.new(<name>, <method>, map._is_virtual)`) without the file being present on the local file system.

As this happens in the init phase of HAProxy I believe there should be no impact on the configuration of the rest of HAProxy. The `map._is_virtual` flag (or something else) can indicate to other parts of HAProxy that this is a non-file map if needed.

### What are you trying to do?

Currently HAProxy requires a pattern file to exists in the local file system (even when it's empty). For this use case we're retrieving information from an external system and put it in a local map in Lua.

the `core.register_init()` function is used to create the map instance and a `core.register_task()` instance is tasked with keeping this map up to date.

The use case here is for example that the external system contains the information needed to block an IP address and when HAProxy is deployed automatically it can find this information from a central external service instead of an external system having to talk to HAProxy to update the map.

Having to always make sure the (empty) file exist makes it more difficult to deploy HAProxy in a more dynamic environment. 

### Output of `haproxy -vv`

```plain
HAProxy version 2.8.0-fdd8154 2023/05/31 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.0.html
Running on: Linux 5.15.49-linuxkit-pr #1 SMP PREEMPT Thu May 25 07:27:39 UTC 2023 aarch64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=5).
Built with OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
Running on OpenSSL version : OpenSSL 1.1.1n  15 Mar 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.36 2020-12-04
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.2.1 20210110

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [BWLIM] bwlim-in
        [BWLIM] bwlim-out
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

**Comment by TheDJVG** [comment]
~~Slight oversight on my part is that it seems that currently there's no way to alter the Map from LUA. It seems that this is only possible for ACLs (`core.add/del_acl()`). This feature request only really makes sense if we're able to update the map from LUA too.~~

~~Currently only `core.del_map()` exists.~~

Updating the map can be achieved using `core.set_map()`. I just didn't the documentation well enough.

**Comment by capflam** [comment]
The same function is used to load a map file from a  lua  script and  from anywhere in the configuration. Thus the feature request should not be specific to the lua. And I guess it should be extended to the ACLs file because we can add/del ACL dynamically. 

Technically speaking, it is possible to fallback on an "virtual map" if no file is found and if the map name contains a specific prefix. But It may be  error-prone, So such prefix must be carefully chosen.

Functionally speaking, using map or pattern lists not based on files means their content will be lost on reload, just like any update. If such feature is added, we can bet another one requesting a sync on reload will be created. I wonder if there is a real demand for this functionality apart from yours. It does not means it is not a legitimate use-case. But it might be good to have the option of other users.

**Comment by TheDJVG** [comment]
The main reason for this request is the fact that HAProxy will happily load an empty file but it still requires the file to be there.

Maybe a global option would work that allows the "loading" of non-existing patterns?

Regarding the reload of maps, based on other issues/requests I've seen it's clear HAProxy will never write this data back (unless it's the state file for backends) so I think people are already aware changes made to Maps/ACLs during runtime do not survive.

To add bit more colour to the request is that for this use case some information is stored in an external system that HAProxy queries versus a system that has to contact every instance of HAProxy/dataplane API to make changes to the maps.

**Comment by TheDJVG** [comment]
Created a pull request to add this functionality: https://github.com/haproxy/haproxy/pull/2209

Feel free to let me know if you're interested to pull this or if it needs more attention. Thanks!

**Comment by wtarreau** [comment]
Hello,

it's interesting that you're facing this need, because we've discussed a few times about this over time and even very recently. This needs a bit more discussion actually because other needs were identified which overlap a lot.

First, I do agree wiht you that sometimes it would be desirable to be able to have empty maps. I've felt this need several times when trying to store classification of a client or something like this (e.g. tag abusers).
Of course sometimes it can be done using stick-tables but sometimes it's also convenient via maps or ACLs. Other use cases could include excuse pages for URLs currently under maintenance, that are not used by default and that are directly remapped and used as-is when a URL is matched.

Second, we found ourselves several times facing the situation where a map file had to be created just to store 2-3 entries. When this happens, some might prefer to be able to use the main config file with a dedicated section to fill that map. But the problem then is the syntax: while it can seem trivial to store IP addresses and backend names that will not get mangled by the config parser, placing some regex containing "$" signs, or having spaces in the patterns will quickly become annoying. Also, what to do with leading spaces used for indenting ? How to spot the end without risking to mix entries with config keywords ? Due to this, for now this has been left to rot for several years. But such a principle could have worked by allowing you to declare empty maps using a single line.

Christopher mentioned that one possibility to at least allow to use empty map files could be to use a special syntax for their names. For example we're using "abns@blah" for abstract namespace sockets and even other formats when needed, maybe we could just create a special prefix for map file paths that would not conflict with existing deployments, and would explicitly designate the map file as virtual (this will basically just skip the open/read/close sequence and let the rest as-is).

It seems that this last approach would be the closest to what we have, yet the most extensible (we could imagine pre-fill such maps from the config later). Maybe you (or others) have alternate ideas ?

**Comment by unixsurfer** [comment]
IMHO I don't see the value to declare an `empty` file with some special syntax.

I do understand and acknowledge the fact HAProxy users have been _trained_ to never set a ACL/MAP file that doesn't exist, so you may want to _force_ them to use a special syntax for this case to make them aware about the change in semantics.

But the proposed patch doesn't change that expected behavior as `tune.pattern.allow-non-existing-file` defaults to `0`, so users don't even notice the new functionality and the change in the behavior.

Having said that, as an operator I fill a bit strange to see a configuration file which references other files that don't exist. It is not uncommon but it still feels weird.

**Comment by wtarreau** [comment]
Hi Pavlos! I can understand the need for a virtual map file, because maps permit to do a lot of things, they're just associative arrays and can have plenty of use cases. However, they're referenced by the file name right now, so we could very well imagine to support an optional path prefix that serves to set the namespace. For example HTTP started with a pure file path and finally grew to support a scheme and authority in front of it. Here it's the same, we could imagine that names starting with "file:" are always files and that "virtual:foobar" is a virtual map. That's just an example of course. We do already support things like this for addresses used in logs or servers for example. And that's also why I'd prefer to see this as a prefix rather than as a "file that does not exist", because I'm much more concerned that it will silently ignore typos in the config.

Also we can have other prefixes indicating that a file optionally exists, and that could simplify some setups allowing to preset file names and create empty maps if they're not there, thus avoiding any error or the user being forced to issue a "touch" to create an empty file.

**Comment by unixsurfer** [comment]
> Hi Pavlos! I can understand the need for a virtual map file, because maps permit to do a lot of things, they're just associative arrays and can have plenty of use cases. However, they're referenced by the file name right now, so we could very well imagine to support an optional path prefix that serves to set the namespace. For example HTTP started with a pure file path and finally grew to support a scheme and authority in front of it. Here it's the same, we could imagine that names starting with "file:" are always files and that "virtual:foobar" is a virtual map. That's just an example of course. We do already support things like this for addresses used in logs or servers for example. And that's also why I'd prefer to see this as a prefix rather than as a "file that does not exist", because I'm much more concerned that it will silently ignore typos in the config.
> 

I now see the value, thanks for the very detailed explanation. It makes a lot of sense having  the prefix `virtual:foobar` .
+1 from me then.

**Comment by capflam** [comment]
About this feature, it could be good to decide how it should work. What kind of prefix ? We should support a prefix for files (the default behavior when there is no prefix) and a prefix for the "virtual" maps. In this last case, the map name is just an id. Should we add a third prefix to operationally load a map from a file and silently fallback to a virtual map if no file is found ?

About the prefix name, we can use:
 *  `file` for files 
 * `virtual` or `id` for virtual file
 *  and eventually `opt-file` for optional file (not really good but I have no other idea)

About the separator, it may a `:` or `@`. The last one is already used as prefix separator while the `:` is used to split parameters (mainly, `address;port` and `user:passwd` )

**Comment by aderumier** [comment]
Hi,
Sorry for the noise,  but I think it could be great to be able to fill a full map in batch,
instead of setting key by key with core.set_map  (which need some msleep between each insert).

One usecase is for crowdsec
https://github.com/crowdsecurity/cs-haproxy-bouncer

where a blacklist of ip is retrieve from an external api and then filled with set_map, key by key.

**Comment by wtarreau** [comment]
Actually a batched load would be a loop over individual patterns, because each has to be parsed, indexed etc. Look for example how the CLI "add-map" command works. It supports a payload for the argument, which contains series of key+values which are fed one at a time. The trick is that when doing bulk insert, you normally just have to set a new generation number so that new entries will not be visible until atomically committed. To be honest I'm not fluent enough in Lua to know if we have the exact same possibilities, but the internal API and the CLI API already supports this, so maybe it's just a matter of adding some functions (e.g. add_map instead of set_map ?). I'm seeing a `core.set_map`, `Map.new` etc, I'm not sure how all of this is used together, but I feel like you might be able to have a closer look, so do not hesitate to propose something ;-)

**Comment by aderumier** [comment]
yes, I was thinking about porting something like "add_map" && "clear_map", already available in runtime api.
and be able to send an array of ips instead reading a real map file.

The problem with crowsec, with set_map /del_map, it's that when you have only new 20000 ips to update, the loop can take 30s (with a msleep(1) between each ip), and it use almost 1core.   That's works fine for small incremental update,  but with a big update, I think I'll be faster to fully clear/reload the full map.

But I'm a pretty poor C developper to implement this ^_^

**Comment by wtarreau** [comment]
I don't understand where this "msleep(1)" you mentioned twice comes from. Last I checked it was possible to create 200k or so map entries per second and I'm very confused about this point.

Also regarding being a "pretty poor C developer", that's what most contributors used to say before providing useful features :-)  We're still here to help and no one will judge your code. It turns out that aside a few traps C is particuarly easy to re-learn because there are very few things to know, so do not hesitate to try to experiment by modifying the set_map Lua binding to try to use add_map instead and see what you can come up with. It could very well be much easier than you think, and if you need help to finish it will be easier for everyone involved.

**Comment by aderumier** [comment]
sorry, about the msleep, it's in crowsec lua code
https://github.com/crowdsecurity/cs-haproxy-bouncer/commit/bd547b0b988d49f3f06420ca710ec77093a45a6e

without it, calling add_map in loop, is hanging haproxy, because of the lock here:

https://github.com/haproxy/haproxy/blob/master/src/hlua.c

```
static int hlua_set_map(lua_State *L)
{
...
HA_RWLOCK_WRLOCK(PATREF_LOCK, &ref->lock);
	if (pat_ref_find_elt(ref, key) != NULL)
		pat_ref_set(ref, key, value, NULL, NULL);
	else
		pat_ref_add(ref, key, value, NULL);
HA_RWLOCK_WRUNLOCK(PATREF_LOCK, &ref->lock);
}
```
(Note that it was a spinlock previously , and my test are done one 2.6, I don't known if HA_RWLOCK_WRLOCK is faster ? )
https://github.com/haproxy/haproxy/commit/81815a9a83517e0b4f2ec0e33f5de02e3927d86c)



I'll look to add the lua bind for add_map, the set_map bind don't see too complex.

**Comment by wtarreau** [comment]
There's absolutely no reason a sleep could avoid the effect of a lock. What I'm suspecting instead is a matter of processing time instead: inserting too many elements at once using set_map() used to take a huge amount of time due to the linear search of the old elements to update/replace. Normally it's sufficient to call core.yield() to avoid this, it gives the CPU back to haproxy for other tasks and comes back here ASAP. And in 2.9 the lookup is much faster so this would be a problem of the past. However, doing too much at once still continues to inflict large processing delays, and it's a good idea to still have a yield every N insertions (N to be determined, if you see that every 1000 starts to show increased response times, do it every 10 and that should be good).

**Comment by aderumier** [comment]
Thanks Willy !  I'll try to test 2.9 to compare, and try to implement core.yield.
(and I'll try to send a patch for add_map|clear_map this month)

**Comment by aderumier** [comment]
(BTW, in 2.6, set_map seem quite slow indeed,  without any sleep, I'm around 1s for 20000 set_map).

(Edit: it's 1s for 20000 set_map, not 10s)

**Comment by capflam** [comment]
FYI, it is planned for the 3.0 (already in the 2.9-next). It was not pushed in the 2.9 because it is freezed. The 2.9.0 is planned for tomorrow.

**Comment by capflam** [comment]
Fixed now in 3.0-DEV. Thanks !
